### PR TITLE
Create an GH Actions workflow covering testing of Handsontable and the wrappers.

### PR DESCRIPTION
**Pull request by jansiegel**
### Context
This PR adds a test workflow, covering:
- Handsontable E2E, Walkontable, unit and type tests
- All wrapper tests

Additional changes:
- I needed to make the angular wrapper's tests run on headless Chrome on CI, because of the issues with GHA.
- I split some scripts in `package.json` to smaller ones, to reuse parts of them in the workflow.
- I made the jasmine tests display the spec "dots" display in same-lenghed chunks, because of the issue with GHA's log preview (it crashed the browser)

TODO:
- Possibly force the workflow triggered on `push` events to compare changes against `develop` by default? It seems like the auto-recognized merge-base is not too reliable.
- The Handsontable's change paths are temporarily set as `'src/**'` - this should be changed after moving Handsontable to its own separate directory.

### How has this been tested?
The workflow should get triggered on `push` and a subset of `pull_request` events. If a `pull_request` event fires, the doubled workflow triggered on `push` should skip all jobs.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #8652 

### Affected project(s):
- [x] `handsontable`
- [x] `@handsontable/angular`
- [x] `@handsontable/react`
- [x] `@handsontable/vue`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [ ] My change requires a change to the documentation.

### CODE REVIEW

**Comment by wojciechczerniak** [comment]
Screenshot 2021-09-20 at 16 12 57

Cool. Angular tests are run because you've changed it's package.json nad karma.conf.
Handsontable e2e tests are run because of config change? We have 2x 10min here. Did we win something here?

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+        uses: actions/download-artifact@3be87be14a055c47b01d3bd88f8fe02320a9bb60 # https://github.com/actions/download-artifact/releases/tag/v2.0.10
+        with:
+          name: handsontable-build-es-cjs
+          path: ./
+      - run: tar -zxf tmp.tar.gz ./tmp && rm tmp.tar.gz
+      - run: npm run swap-package-links
+      - run: npm run test:unit
+
+  test-handsontable-types:
+    name: "[TEST] Handsontable: types"
```
Comment (wojciechczerniak) [inline .github/workflows/test.yml:191]:
```suggestion
    name: "[TEST] Handsontable: Types"
```

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@25316bbc1f10ac9d8798711f44914b1cf3c4e954 # https://github.com/actions/setup-node/releases/tag/v2.4.0
+        with:
+          node-version: 15.9.0
+          cache: 'npm'
+      - run: npm ci --no-audit
+      - run: npm run test:walkontable
+
+  test-handsontable-e2e:
+    name: "[TEST] Handsontable: UMD"
```
Comment (wojciechczerniak) [inline .github/workflows/test.yml:234]: Should we call it E2E? It's using UMD build but it's testing e2e for all builds, right?

**Review by budnix — COMMENTED** [review]
It looks very impressive! But is there is a change to optimize the job that does the same? I mean when I created a [PR](https://github.com/handsontable/handsontable/pull/8732) I've spotted that half of the jobs do the same.

_Edit: Similar problem https://github.community/t/how-to-trigger-an-action-on-push-or-pull-request-but-not-both/16662._

2021-09-21 o 09 52 22

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,357 @@
+name: Test all packages
+
+on:
+  pull_request:
+    types:
+      - synchronize
+      - opened
+      - reopened
```
Comment (budnix) [inline .github/workflows/test.yml:8]: There is a question. Do we need trigger tests on PR open, reopen? I mean these tests are strictly related to the source code change 🤔

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,357 @@
+name: Test all packages
```
Comment (budnix) [inline .github/workflows/test.yml:1]: Can we all these jobs as "Required"?

File: .github/workflows/test.yml (hunk 2 above)
Comment (jansiegel) [reply .github/workflows/test.yml:234]: I think it's using only the UMD build from `/dist/handsontable.js`.

File: .github/workflows/test.yml (hunk 3 above)
Comment (jansiegel) [reply .github/workflows/test.yml:8]: You're probably right 👍

**Comment by budnix** [comment]
I've tested the GA Workflow on a separate [branch](https://github.com/handsontable/handsontable/pull/8732) and I've got some thoughts:
 * The Handsontable is built and tested unnecessarily when I changed only the source code related to the docs (https://github.com/handsontable/handsontable/pull/8732/commits/6848202f4254e26d3d8ded7491a748ed8ab8dcfb);
 * As above the same when I change only docs MD file (https://github.com/handsontable/handsontable/pull/8732/commits/f5c9f77572909f16c9c999d1a8c5390e8a8c81cd);
 * The Handsontable is built only when the Examples code is touched (https://github.com/handsontable/handsontable/pull/8732/commits/51748b0532cdce21c71068fa2281af669d918aad);

Analyzing these findings I came up to the conclusion that the tests are not triggered by each commit change separately but the changes in the source code are stacked and the commits only trigger the jobs. @jansiegel am I right or missing something?

**Comment by jansiegel** [comment]
> It looks very impressive! But is there is a change to optimize the job that does the same? I mean when I created a [PR](https://github.com/handsontable/handsontable/pull/8732) I've spotted that half of the jobs do the same.
> 
> _Edit: Similar problem https://github.community/t/how-to-trigger-an-action-on-push-or-pull-request-but-not-both/16662._
> 
> 2021-09-21 o 09 52 22

I tried to create one workflow for all scenarios, so it runs on both `push` and `pull_request`, to handle situations, where a branch doesn't have a PR (yet). However, if there _is_ a PR created, all the jobs triggered by `push` should be skipped (but they seem to be marked as "completed" here, so maybe something's not working right..?). I'm not a huge fan of having them all cluttering the "checks" box in the issues view, but I don't know if they can be hidden somehow?

**Comment by jansiegel** [comment]
> I've tested the GA Workflow on a separate [branch](https://github.com/handsontable/handsontable/pull/8732) and I've got some thoughts:
> 
> * The Handsontable is built and tested unnecessarily when I changed only the source code related to the docs ([6848202](https://github.com/handsontable/handsontable/commit/6848202f4254e26d3d8ded7491a748ed8ab8dcfb));
> * As above the same when I change only docs MD file ([f5c9f77](https://github.com/handsontable/handsontable/commit/f5c9f77572909f16c9c999d1a8c5390e8a8c81cd));
> * The Handsontable is built only when the Examples code is touched ([51748b0](https://github.com/handsontable/handsontable/commit/51748b0532cdce21c71068fa2281af669d918aad));
> 
> Analyzing these findings I came up to the conclusion that the tests are not triggered by each commit change separately but the changes in the source code are stacked and the commits only trigger the jobs. @jansiegel am I right or missing something?

The changes are indeed stacked and the whole chunk of changes in the PR is being analyzed on the `pull_request` trigger. I _think_ it makes sense in most cases, as the PR is what is going to be eventually merged into the base branch, not a single commit. It is counter-intuitive when testing and trying out different things, though.

**Comment by jansiegel** [comment]
> Screenshot 2021-09-20 at 16 12 57
> 
> Cool. Angular tests are run because you've changed it's package.json nad karma.conf.
> Handsontable e2e tests are run because of config change? We have 2x 10min here. Did we win something here?

These are run because of the change in `test/scripts/run-puppeteer.js`. I wasn't sure it's supposed to work that way, but I figured changes in `test/scripts/**` can affect the way tests are being performed, so I added it to the path whitelist.

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+              - 'scripts/**'
+              - 'test/unit/**'
+            test-handsontable-types:
+              - '.config/**'
+              - 'src/**'
+              - 'scripts/**'
+              - 'test/types/**'
+            test-handsontable-walkontable:
+              - '.config/**'
+              - 'src/**'
```
Comment (budnix) [inline .github/workflows/test.yml:77]:
```suggestion
              - 'src/3rdparty/walkontable/**'
```

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+              - 'scripts/**'
+              - 'test/scripts/**'
```
Comment (budnix) [inline .github/workflows/test.yml:79]:
```suggestion
              - '.config/**'
              - 'src/3rdparty/walkontable/**'
```

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+    if: |
+      needs.check-scope.outputs.run-all == 'true' ||
+      (needs.check-scope.outputs.skip-all == 'false' &&
+      needs.check-scope.outputs.build-handsontable-umd == 'true')
+    steps:
+      - uses: actions/checkout@5a4ac9002d0be2fb38bd78e4b4dbde5606d7042f # https://github.com/actions/checkout/releases/tag/v2.3.4
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@25316bbc1f10ac9d8798711f44914b1cf3c4e954 # https://github.com/actions/setup-node/releases/tag/v2.4.0
+        with:
+          node-version: 15.9.0
```
Comment (budnix) [inline .github/workflows/test.yml:121]: WDYT about extracting the node version as env ([docs](https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#example-setting-an-environment-variable))? The node version is repeated 10 times right now.

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+
+on:
+  pull_request:
+    types:
+      - synchronize
+  push:
+    branches:
+      - '**'
+
+jobs:
```
Comment (budnix) [inline .github/workflows/test.yml:11]: I wonder if we can cache the `npm i` as mentioned [here](https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows#example-using-the-cache-action) for all below jobs. I'm curios how much time we can safe 🤔

File: .github/workflows/test.yml (hunk 6 above)
Comment (jansiegel) [reply .github/workflows/test.yml:79]: Are you sure we can remove these two? `test/scripts/**` can potentially change the way `puppeteer` runs tests, right?

Comment (budnix) [reply .github/workflows/test.yml:79]: Right, the `test/scripts/**` entry can be helpful here.

File: .github/workflows/test.yml (hunk 8 above)
Comment (jansiegel) [reply .github/workflows/test.yml:11]: I think the `setup-node` action we're using already takes care of that:

> The action has a built-in functionality for caching and restoring dependencies. It uses actions/cache under hood for caching dependencies but requires less configuration settings. Supported package managers are npm, yarn, pnpm (v6.10+). The cache input is optional, and caching is turned off by default.

https://github.com/actions/setup-node#caching-packages-dependencies

Comment (budnix) [reply .github/workflows/test.yml:11]: Thanks for checking that 🎖️

**Comment by budnix** [comment]
@jansiegel and last thing. Don't forget to update the README.md file. The Codeship badge has to be replaced with something from GHA I suppose.

**Review by budnix — CHANGES_REQUESTED** [review]
Everything looks great 👏   but when I create the PR with some changes that should trigger the workflow the jobs do nothing https://github.com/handsontable/handsontable/pull/8738. The next commit push to the branch works well.

**Comment by jansiegel** [comment]
@budnix 

> Everything looks great 👏 but when I create the PR with some changes that should trigger the workflow the jobs do nothing #8738. With the next commit pushed to the branch, it works well.

I guess it's due to the fact I removed `open` and `reopen` from the `pull_request` triggers https://github.com/handsontable/handsontable/pull/8653#discussion_r712806043. That's a good use case to bring them back, what do you think?

**Comment by budnix** [comment]
> @budnix
> 
> > Everything looks great 👏 but when I create the PR with some changes that should trigger the workflow the jobs do nothing #8738. With the next commit pushed to the branch, it works well.
> 
> I guess it's due to the fact I removed `open` and `reopen` from the `pull_request` triggers [#8653 (comment)](https://github.com/handsontable/handsontable/pull/8653#discussion_r712806043). That's a good use case to bring them back, what do you think?

If that would solve the problem I am for trying that.

**Comment by jansiegel** [comment]
@budnix 
> If that would solve the problem I am for trying that.

I reverted this change in https://github.com/handsontable/handsontable/pull/8653/commits/233280cde742f6254e320d556e9e4541da3fda1c, does it work correctly now?

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+      - synchronize
+  push:
+    branches:
+      - 'master'
+      - 'develop'
+      - 'release/**'
+  workflow_dispatch:
+
+env:
+  NODE_VERSION: 15.9.0
```
Comment (budnix) [inline .github/workflows/test.yml:17]: I see that our all workflow files use `'15.x'` ([changelog.yml](https://github.com/handsontable/handsontable/blob/feature/issue-8652/.github/workflows/changelog.yml#L27), [linter.yml](https://github.com/handsontable/handsontable/blob/feature/issue-8652/.github/workflows/linter.yml#L15) etc.). WDYT to do the same here?

**Review by budnix — APPROVED** [review]
🏅

**Review by wojciechczerniak — APPROVED** [review]
second that! 👏

**Pull request merged**